packages/gatsby-transformer-video/src/ffmpeg.js

```diff
@@ -0,0 +1,443 @@
+      maxHeight,
+      duration,
+      fps,
+      saturation,
+      overlay,
+      overlayX,
+      overlayY,
+      overlayPadding,
+    } = fieldArgs
+    const filters = []
+    const { duration: sourceDuration } = info.streams[0]
+
+    if (duration) {
+      filters.push(`setpts=${(duration / sourceDuration).toFixed(6)}*PTS`)
+    }
+
+    if (fps) {
+      filters.push(`fps=${fps}`)
+    }
+
+    if (maxWidth || maxHeight) {
+      filters.push(`scale=${this.generateScaleFilter({ maxWidth, maxHeight })}`)
+    }
+
+    if (saturation !== 1) {
+      filters.push(`eq=saturation=${saturation}`)
+    }
+
+    if (overlay) {
+      const padding = overlayPadding === undefined ? 10 : overlayPadding
+      let x = overlayX === undefined ? `center` : overlayX
+      let y = overlayY === undefined ? `center` : overlayY
+
+      if (x === `start`) {
+        x = padding
+      }
+      if (x === `center`) {
+        x = `(main_w-overlay_w)/2`
+      }
+      if (x === `end`) {
+        x = `main_w-overlay_w-${padding}`
+      }
+
+      if (y === `start`) {
+        y = padding
+      }
+      if (y === `center`) {
+        y = `(main_h-overlay_h)/2`
+      }
+      if (y === `end`) {
+        y = `main_h-overlay_h-${padding}`
+      }
+
+      filters.push(`overlay=x=${x}:y=${y}`)
+    }
+
+    return filters
+  }
+
+  // Apply required changes from some filters to the fluent-ffmpeg session
+  enhanceFfmpegForFilters = ({
+    fieldArgs: { overlay, duration },
+    ffmpegSession,
+  }) => {
+    if (duration) {
+      ffmpegSession.duration(duration).noAudio()
+    }
+    if (overlay) {
+      const path = resolve(this.rootDir, overlay)
+      ffmpegSession.input(path)
+    }
+  }
+
+  createH264 = (...args) => this.queue.add(() => this.convertToH264(...args))
+  convertToH264 = async ({ publicDir, path, filename, fieldArgs, info }) => {
+    const { crf, preset, maxRate, bufSize } = fieldArgs
+
+    const publicPath = resolve(publicDir, `${filename}-h264.mp4`)
+
+    const alreadyExists = await pathExists(publicPath)
+    if (!alreadyExists) {
+      console.log(`[H264] Converting ${path} (${filename})`)
+
+      const filters = this.createFilters({ fieldArgs, info }).join(`,`)
+
+      console.log(`Applied complex filter: ${filters}`)
+
+      const currentFps = parseInt(
+        info.streams
+          .find(stream => stream.codec_type === `video`)
+          .r_frame_rate.split(`/`)[0]
+      )
+
+      const outputOptions = [
+        crf && `-crf ${crf}`,
+        preset && `-preset ${preset}`,
+        !crf && maxRate && `-maxrate ${maxRate}`,
+        !crf && bufSize && `-bufsize ${bufSize}`,
+        `-movflags +faststart`,
+        `-profile:v high`,
+        `-bf 2	`,
+        `-g ${Math.floor((fieldArgs.fps || currentFps) / 2)}`,
+        `-coder 1`,
+        `-pix_fmt yuv420p`,
+      ].filter(Boolean)
+
+      console.log(
+        `ffmpeg command:\n\nffmpeg -i ${path} -vf "${filters}" -c:v libx264 ${outputOptions.join(
```

Comment:
- for all `ffmpeg command` logs: i would avoid "guessing" the resulting ffmpeg command, this info can not be always in sync with the generated command (if you forget to change it or if `fluent-ffmpeg` changed internally) and gives you wrong debug informations (for example it not shows the changes from `enhanceFfmpegForFilters`)
- i think it is better to use the "real" debug info from fluent-ffmpeg with `.on('start', function(commandLine) { console.log(commandLine);})`
- see my comment at `executeFfmpeg`

Reply:
Thanks, I took this into account in the last refactor :)

Reply:
done